### DISCUSSION TITLE
Section links and named destinations

### DISCUSSION THREAD

**Discussion by lgsndbox** [QuestPDF/QuestPDF#180] [Q&A] [answered]
Hello,

I need to add a "table of contents" (TOC) section to a PDF document consisting of merged PDFs. The TOC should have (using PDF terminology) Annotations of the type Link whose Destinations refer to the first page of each of the merged PDFs.

I wonder if there would be a way to achieve this using PDFQuest.
From what I understand, there is no way to add pages from an existing PDF document to a QuestPDF Document.
It implies that I'll have to go thru the following steps:
- Create the TOC using QuestPDFand render it as a PDF.
- Merge this in the first position with the already merged PDFs.
Using another PDF library, such as PDFSharp will be necessary for the merging part.

To have working hyperlinks in my TOC, I see two options:
- Find the location of the "soon-to-be link" text in the TOC, calculate the rectangle around it, and create an Annotation of type Link that will refer to the correct Destination. Not easy!
- Have Named Destinations in the merged PDF document (before merging with the TOC), and add the corresponding names to QuestPDF Document Hyperlinks, even if they do not exist in the QuestPDF Document.

Would that second solution work?
When you create a QuestPDF SectionLink, does that produce an Annotation of type Link whose /Dest value type is a name (and not an explicit Destination, such as [22 0 R /Fit]?

Or am I completely off, and another way exists in order to achieve this?

Thanks a lot.

**Comment by lgsndbox** [comment, marked as answer]
To answer:

"When you create a QuestPDF SectionLink, does that produce an Annotation of type Link whose /Dest value type is a name (and not an explicit Destination, such as [22 0 R /Fit]?"

I simply tested it.

Indeed, adding a section creates a Named Destination, and the referring SectionLink creates an Annotation of type Link that uses as the /Dest value the name of the section. This is good!

Is this behavior guaranteed? or is it an implementation detail that may change?

If this behavior is guaranteed, it should be possible to use Named Destinations not created by QuestPDF, and have QuestPDF SectionLink use them, given the corresponding Name is known at the QuestPDF Document creation time.

Is there any better solution, or issues I did not foresee?

**Reply by MarcinZiabek** [reply]
> Is this behavior guaranteed? or is it an implementation detail that may change?

Right now the implementation is based on the default behavior of SkiaSharp. I want to preserve named destinations and named links. They are simply more predictable and easier to use. So there are no plans to change this functionality and it can be considereed as guaranteed.

> If this behavior is guaranteed, it should be possible to use Named Destinations not created by QuestPDF, and have QuestPDF SectionLink use them, given the corresponding Name is known at the QuestPDF Document creation time.

Yes, I think so. Altough, I have never merged QuestPDF output with other PDF files. 

> Is there any better solution, or issues I did not foresee?

I am not aware about any but open to discuss and answer your questions 😁

**Reply by lgsndbox** [reply]
Thanks a lot for your answer!

P.S. I don't know how to mark your answer as the answer...

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
